### PR TITLE
exo test: remove dependency on service role

### LINKED ISSUE #650: exo test: dependecy on service role
exo test depends on the fact that the service name / role (key in app yaml) matches the service directory. Remove this requirement as some services may be used multiple types in an applicaiton.

`exo test` should iterate through the locations (not the keys)

### PR DESCRIPTION
**Pull request by charlierudolph**
resolves #650

### CODE REVIEW

File: src/application/tester.go [hunk 1]
```diff
@@ -17,7 +17,12 @@ func TestApp(appContext types.AppContext, logger *util.Logger, mode composebuild
 	}
 
 	numFailed := 0
+	locations := []string{}
 	for serviceName, serviceContext := range serviceContexts {
+		if util.DoesStringArrayContain(locations, serviceContext.Location) {
+			continue
+		}
+		locations = append(locations, serviceContext.Location)
```
Comment (charlierudolph) [inline src/application/tester.go]: This is preventing from running the same tests twice when a service is reused. Wasn't exactly certain how to test this

File: src/application/service_tester.go [hunk 2]
```diff
@@ -88,11 +88,13 @@ func (s *ServiceTester) getDockerComposeConfig() (types.DockerCompose, error) {
 	if err != nil {
 		return result, err
 	}
-	dockerConfigs := types.DockerConfigs{}
-	serviceDockerConfig := appDockerConfigs[s.Role]
-	serviceDockerConfig.DependsOn = s.getDependencyContainerNames()
-	serviceDockerConfig.Command = s.ServiceConfig.Development.Scripts["test"]
-	dockerConfigs[s.Role] = serviceDockerConfig
+	dockerConfigs, err := s.Initializer.getServiceDockerConfig(s.Role, s.ServiceConfig, types.ServiceData{Location: s.Role})
```
Comment (hugoesthere) [inline src/application/service_tester.go]: Why would `Location` be `s.Role` here?

File: src/application/service_tester.go [hunk 3]
```diff
@@ -88,11 +88,13 @@ func (s *ServiceTester) getDockerComposeConfig() (types.DockerCompose, error) {
...
+	if err != nil {
+		return result, err
+	}
+	serviceConfig := dockerConfigs[s.Role]
+	serviceConfig.DependsOn = s.getDependencyContainerNames()
+	serviceConfig.Command = s.ServiceConfig.Development.Scripts["test"]
```
Comment (hugoesthere) [inline src/application/service_tester.go]: It doesn't look like these `serviceConfig` objects actually being used anywhere, beyond being assigned values here.

File: src/application/tester.go (hunk 1 above)
Comment (hugoesthere) [reply src/application/tester.go]: This seems simple and sound enough. If you want, you could extract a `forServiceDirs` method like we have `forEachService` in the `types/app-config.go` that just loops through service locations and maybe test that method. We might be using something like that in `exo clean` as well you said?

File: src/application/service_tester.go (hunk 2 above)
Comment (charlierudolph) [reply src/application/service_tester.go]: The role in this case is the service dir (we need your rename pr in)

File: src/application/service_tester.go (hunk 3 above)
Comment (charlierudolph) [reply src/application/service_tester.go]: The `serviceConfig` variable is only used to update the command and dependsOn. Its a value in the `dockerConfigs` map which is what is returned

File: src/application/tester.go (hunk 1 above)
Comment (charlierudolph) [reply src/application/tester.go]: That sounds good

File: src/application/service_tester.go (hunk 3 above)
Comment (hugoesthere) [reply src/application/service_tester.go]: It doesn't seem like this is actually modifying the underlying `dockerConfigs` object. I had to reassign `serviceConfig` to `dockerConfigs[s.Role]` in order for this method to return the right objects

**Comment by hugoesthere** [comment]
@charlierudolph should be good for another look

**Pull request merged**